design-proposals/memory-hotplug.md

```diff
@@ -0,0 +1,209 @@
+
+## Extensibility
+
+We will be able to use the `VMLiveUpdateFeatures` section to manage things like cpu/memory hotplug .
+
+## Update/Rollback Compatibility
+
+The feautre doesn't intorduce any breaking changes so existing functionality of versions prior to this feature should
+not be impacted upon update/rollback. However if existing workloads would like to opt-in for this feature after an
+update, the VMI should be re-created since the addition of new fields to the VMI API.
+
+## Functional Testing Approach
+
+Functional tests will simply:
+
+* Create and start a VM with `memeory` set as one of th `liveUpdateFeatures`
+* Modify the number of memory guest for the VM
+* Ensure the guest sees the changed number of memory
+* Ensure the memory resources on the virt-launcher pod were adjusted accordingly
+* Ensure interoperability with non-hotplug related live-migration occasions (workload-update, evacuation,
+  user-triggered)
+
+## Implementation Phases
+
+### Phase 1: Creating the KubeVirt API and linking it to the LibVirt memory hotplug API
+
+### Phase 2: Triggering a live migration when the number of sockets changes, to adjust the virt-launcher memory resources
+
+The
+VMI [workload update controller](https://github.com/kubevirt/kubevirt/blob/ce8aab7874d2e1586787e3e2a17306b7edca1b8a/pkg/virt-controller/watch/workload-updater/workload-updater.go#L476)
+is extended to live migrate VMIs to satisfy updates to a VMI resource request/limits field.
+
+Sequence of events:
+
+* User updates memory guest on a vm.spec.template.spec.domain.memory
+* VM controller determines the request/limits required and writes the necessary changes to the active vmi.Spec
+* VM controller writes the required request/limits in the vmi.Status
+* VMI workload update controller detects the mismatch between the resource request/limits on the pod vs vmi.Spec and
+  live migrates the VMI to satisfy vmi.spec
+* Migration controller constructs a new target pod with the desired memory request/limits
+* Migration controller identfies a match between the required request/limits on the vmi.Status and those of the VMI
+  target pod
+* Migration controller marks the VMI as safe to hotplug.
+* VMI live migrates.
+* virt-handler identifies the mark and triggers the hotplug/hot-unplug LibVirt action after the live migration completes
+* VM status is updated to reflect that the pending hotplug/hot-unplug actions have completed
+
+### Phase 3 (future): Using the [Kubernetes vertical scaling API](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1287-in-place-update-pod-resources) once it reaches a more mature state
+
+The inplace vertical pod feature allows us to dynamically change the VMI pods resource request/limits. This feature will
+allow us to avoid live migration, however the issue for us is timing. This feature is alpha in Kubernetes 1.27. We need
+it to be at least in beta to integrate it without risking stability and API reliability.
+
+Sequence of events:
```

Comment:
I would remove the sequence of events here, it's going to be exactly the same as the previous section minus live migration, also it's more related to the implementation itself.

Reply:
remove it